src/passes/OptimizeCasts.cpp

```diff
@@ -102,6 +121,235 @@ namespace wasm {
+      for (size_t i = 0; i < numLocals; i++) {
+        earliestRefAsReachable[i] = nullptr;
+      }
+    }
+  }
+
+  void visitLocalSet(LocalSet* curr) {
+    visitExpression(curr);
+    earliestRefCastReachable[curr->index] = nullptr;
+    earliestRefAsReachable[curr->index] = nullptr;
+  }
+
+  void visitLocalGet(LocalGet* curr) {
+    visitExpression(curr);
+
+    if (!earliestRefCastReachable[curr->index]) {
+      earliestRefCastReachable[curr->index] = curr;
+    }
+
+    if (!earliestRefAsReachable[curr->index]) {
+      earliestRefAsReachable[curr->index] = curr;
+    }
+  }
+
+  void visitRefAs(RefAs* curr) {
+    visitExpression(curr);
+
+    // Right now, since all other ops are not casts, we only care about
+    // RefAsNonNull If more cast ops are added, the program must be changed to
+    // accomodate more than one type of op.
+    if (curr->op != RefAsNonNull) {
+      return;
+    }
+
+    auto* fallthrough = Properties::getFallthrough(curr, options, *getModule());
+
+    if (auto* get = fallthrough->dynCast<LocalGet>()) {
+      auto castIter = refAsToApply.find(earliestRefAsReachable[get->index]);
+      if (castIter == refAsToApply.end()) {
+        refAsToApply.emplace(earliestRefAsReachable[get->index],
+                             get == earliestRefAsReachable[get->index]);
+      }
+    }
+  }
+
+  void visitRefCast(RefCast* curr) {
+    visitExpression(curr);
+
+    // Using fallthroughs here is fine due to the following cases.
+    // Suppose we have types $A->$B->$C (where $C is the most refined)
+    // and $D, which is an unrelated type.
+    // Case 1:
+    // (ref.cast $A (ref.cast $C (local.get $x)))
+    //
+    // ref.cast $C is initially chosen for $x. Then we consider ref.cast $A.
+    // Since $A is less refined than $C, we ignore it.
+    //
+    // Case 2:
+    // (ref.cast $C (ref.cast $A (local.get $x)))
+    //
+    // ref.cast $A is initially chosen for $x. Then we consider ref.cast $C,
+    // which is more refined than ref.cast $A, so we replace it with ref.cast
+    // $C.
+    //
+    // Case 3:
+    // (ref.cast $B (ref.cast $B (local.get $x)))
+    //
+    // We initially choose to move the inner ref.cast $B. When we consider the
+    // outer ref.cast $B, we can see that it has the same type as tge existing
+    // ref.cast $B, so we ignore it.
+    //
+    // Case 4:
+    // (ref.cast $D (ref.cast $C (local.get $x)))
+    //
+    // This would produce a trap and should already be optimized away
+    // beforehand.
+    //
+    // Note that if the best cast is already at the target location, we will
+    // keep it for comparison purposes but remove it once the traversal is done.
+
+    auto* fallthrough = Properties::getFallthrough(curr, options, *getModule());
+    if (auto* get = fallthrough->dynCast<LocalGet>()) {
+      auto castIter = refCastToApply.find(earliestRefCastReachable[get->index]);
+
+      if (castIter == refCastToApply.end()) {
+        RefCastInfo newCast(curr->type,
```

Comment:
@kripken For 3., should I prefer the experimental cast then? Or just ignore it?

Reply:
I guess either way is ok. But it's worth a comment at least mentioning that the ordering between them is arbitrary, just so people reading the code don't wonder "this must be obvious because there's no comment, but I don't know why this is how it is"

Reply:
Removing special treatment of safety for now.